app/scripts/controllers/plugins.js

```diff
@@ -226,68 +286,61 @@ class PluginsController extends EventEmitter {
 
       console.log(`Constructing plugin`)
       plugin = {
-        sourceCode,
+        // manifest: {}, // relevant manifest metadata
+        name: pluginName,
         initialPermissions: namespacedInitialPermissions,
+        permissionName: PLUGIN_PREFIX + pluginName, // so we can easily correlate them
+        sourceCode,
       }
     } catch (err) {
       throw new Error(`Problem loading plugin ${pluginName}: ${err.message}`)
     }
 
+    // always retrieve state close to the call to updateState
+    const pluginsState = this.store.getState().plugins
+
     // restore relevant plugin state if it exists
-    if (pluginState[pluginName]) {
-      plugin = { ...pluginState[pluginName], ...plugin }
+    if (pluginsState[pluginName]) {
+      plugin = { ...pluginsState[pluginName], ...plugin }
     }
-    plugin.pluginName = pluginName
 
     // store the plugin back in state
     this.store.updateState({
       plugins: {
-        ...pluginState,
+        ...pluginsState,
         [pluginName]: plugin,
       },
     })
-  }
 
-  async authorize (pluginName) {
-    console.log(`authorizing ${pluginName}`)
-    const pluginState = this.store.getState().plugins
-    const plugin = pluginState[pluginName]
-    const { sourceCode, initialPermissions } = plugin
-    const ethereumProvider = this.setupProvider({ hostname: pluginName }, async () => {
-      return {name: pluginName }
-    }, true)
+    return plugin
+  }
 
+  authorize (pluginName, ethereumProvider) {
```

Comment:
Well now I'm spoiled and wish I had one of your nice comments on this one. I was lazy and left my plugin lifecycle comments at the top of the file.

Reply:
Hah, this innocuous question really opened pandora's box. We didn't have teardown for plugin `onMetaMaskEvent` listeners! See updated description.